File: Documentation/Lab 7/Lab7_Students_Files/navigation_lab7_line_following_template.py
#!/usr/bin/env python
from __future__ import print_function
import sys
import math
import numpy as np
import time

#ROS Imports
import rospy
from sensor_msgs.msg import LaserScan
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive
from nav_msgs.msg import Odometry




class WallFollow:
    def __init__(self):

        # Read the Wall-Following controller paramters form params.yaml
        # ...
    
        # Subscrbie to LiDAR scan Wheel Odometry topics. This is to read the LiDAR scan data and vehicle actual velocity
        rospy.Subscriber(lidarscan_topic, LaserScan, self.lidar_callback,queue_size=1)
        rospy.Subscriber(odom_topic, Odometry, self.odom_callback,queue_size=1)

        # Create a publisher for the Drive topic
        self.drive_pub =rospy.Publisher(drive_topic, AckermannDriveStamped, queue_size=1)
        
        

     # The LiDAR callback function is where you read LiDAR scan data as it becomes availble and compute the vehile veloicty and steering angle commands
    
    def lidar_callback(self, data):      

      # Exttract the parameters of two walls on the left and right side of the vehicles. Referrring to Fig. 1 in the lab instructions, these are al, bl, thethal, ... 
      # ...
      # Compute the steering angle command to maintain the vehicle in the middle of left and and right walls
      # ...  
 
      # Find the closest obstacle point within a narrow viewing angle in front of the vehicle and compute the vehicle velocity command accordingly
      #  ...     

      # Publish steering angle and velocity commnads to the Drive topic
      # ...


    # The Odometry callback reads the actual vehicle velocity from VESC. 
    
    def odom_callback(self, odom_msg):
        # update current speed
        self.vel = odom_msg.twist.twist.linear.x


def main(args):
    rospy.init_node("WallFollow_node", anonymous=True)
    wf = WallFollow()
    rospy.sleep(0.1)
    rospy.spin()

if __name__=='__main__':
	main(sys.argv)

File: Documentation/Lab 7/Lab7_Students_Files/navigation_lab7_line_following_template.cpp
#define _USE_MATH_DEFINES //M_PI

#include "ros/ros.h"

#include "std_msgs/String.h"
#include "std_msgs/Int32MultiArray.h"
#include "sensor_msgs/LaserScan.h" //receive msgs from lidar
#include "sensor_msgs/Imu.h" //receive msgs from Imu


#include <tf/tf.h> //Quaternions

#include "ackermann_msgs/AckermannDriveStamped.h" //Ackermann Steering

#include "nav_msgs/Odometry.h" //Odometer

#include <string>
#include <vector>


//standard and external
#include <stdio.h>
#include <math.h>
#include <cmath> 
#include <sstream>


class WallFollow 
{
	private:
		ros::NodeHandle nf;

		//Subscriptions
		ros::Subscriber lidar;
		ros::Subscriber odom;
		
		//Publications
		ros::Publisher lidar_pub;
		ros::Publisher driver_pub;	
		
		//topics
		std::string lidarscan_topic, drive_topic, odom_topic;

	public:
		
		WallFollow(){

			nf = ros::NodeHandle("~");
		
        // Read the Wall-Following controller paramters form params.yaml
        //  ...

		//	nf.getParam( ...);

        // Subscrbie to LiDAR scan Wheel Odometry topics. This is to read the LiDAR scan data and vehicle actual velocity
			lidar = nf.subscribe("/scan",1, &WallFollow::lidar_callback, this);
			odom = nf.subscribe(odom_topic,1, &WallFollow::odom_callback, this);

        // Create a publisher for the Drive topic
			driver_pub = nf.advertise<ackermann_msgs::AckermannDriveStamped>(drive_topic, 1);
		}



// The Odometry callback reads the actual vehicle velocity from VESC. 

		void odom_callback(const nav_msgs::OdometryConstPtr& odom_msg){
			vel = odom_msg->twist.twist.linear.x;
		}

     // The LiDAR callback function is where you read LiDAR scan data as it becomes availble and compute the vehile veloicty and steering angle commands


		void lidar_callback(const sensor_msgs::LaserScanConstPtr &data){
			
			
      // Exttract the parameters of two walls on the left and right side of the vehicles. Referrring to Fig. 1 in the lab instructions, these are al, bl, thethal, ... 
      // ...
      // Compute the steering angle command to maintain the vehicle in the middle of left and and right walls
      // ...  
 
      // Find the closest obstacle point within a narrow viewing angle in front of the vehicle and compute the vehicle velocity command accordingly
      //  ...     

      // Publish steering angle and velocity commnads to the Drive topic
      // ...
		}

};



int main(int argc, char **argv){
		ros::init(argc, argv, "navigation");
		WallFollow gb;

		while(ros::ok()){
			ros::spinOnce();
		}

}

